sdv/metadata/table.py

```diff
@@ -205,6 +206,57 @@ def _update_transformer_templates(self, rounding, min_value, max_value):
                 'float': custom_float
             })
 
+    @staticmethod
+    def _get_constraint_sort_key(constraint):
+        try:
+            if not constraint.rebuild_columns:
+                return 0
+            else:
+                return 1
+        except AttributeError:
+            return 1
+
+    def _sort_constraints(self):
+        for idx, constraint in enumerate(self._constraints):
+            if isinstance(constraint, type):
+                constraint = constraint().to_dict()
+            elif isinstance(constraint, Constraint):
+                constraint = constraint.to_dict()
+
+            constraint = Constraint.from_dict(constraint)
+            self._constraints[idx] = constraint
+
+        self._constraints.sort(key=self._get_constraint_sort_key)
+
+    def _validate_constraint_order(self):
```

Comment:
It seems like we are doing the sorting and the validation in two different steps, which all iterate over the constraints multiple times.
Could everything be put in a single loop instead?
Another thing is that the current loop seems to keep track of the individual constraints rather than the columns that they modify. We could probably just keep track of the columns and not the constraints.

So, the logic could be something like:

```python3
@staticmethod
def _prepare_constraints(constraints):
    rebuilt_columns = set()
    transform = []
    reject_sampling = []
    for constraint in constraints:
        if not constraint.rebuild_columns:
            reject_sampling.append(constraint)
        elif rebuild_columns & set(constraint.constraint_columns):
            # previous rebuild columns overlap with current constraint_columns
            # we can recompute the intersection to know which columns overlap
            # and put them on the error message
            raise error
        else:
            transform.append(constraint)
            rebuild_columns.update(constraint.rebuild_columns)

    return reject_sampling + transform
```